File: packages/aarm-sdk/src/frameworks/langchain.ts
import { AarmClient } from '../client';

/**
 * Core configuration overrides for the LangChain session hooks
 */
export interface LangChainHookOptions {
  businessUnit?: string;
  application?: string;
  principalId?: string;
}

/**
 * Creates callbacks/hooks intended to integrate seamlessly with LangChain's Tool execution flow.
 * Note: Actual LangChain types are omitted to avoid steep dependency overhead in the base SDK.
 */
export function createLangchainHooks(client: AarmClient, sessionId: string, intent: string, options?: LangChainHookOptions) {
  return {
    /**
     * Call this before the tool runs to ensure AARM permits the action.
     * @param toolName The name of the tool
     * @param toolInput The parsed input given to the tool
     * @throws Error if the action is DENIED by AARM
     */
    onToolStart: async (toolName: string, toolInput: string | Record<string, unknown>) => {
      let parameters: Record<string, unknown> = {};
      if (typeof toolInput === 'string') {
        try {
          parameters = JSON.parse(toolInput);
        } catch {
          parameters = { input: toolInput };
        }
      } else {
        parameters = toolInput;
      }

      const mediation = await client.mediateAction({
        sessionId,
        intent,
        userPrompt: 'Intercepted tool call', // Ideally pass the exact current run prompt
        actionType: toolName,
        parameters,
        ...options // Passing the dynamic options like principalId
      });

      if (mediation.status === 'DENY') {
        throw new Error(`AARM Policy Violation: Execution of tool '${toolName}' was blocked. Reason: ${mediation.executionStatus}`);
      }

      // 'ALLOW' or 'STEP_UP' (which resolves to ALLOW if successful) continue execution.
      return mediation.receipt;
    },

    /**
     * Can be wired up if post-tool execution telemetry/auditing is needed by AARM in the future
     */
    onToolEnd: (toolName: string, output: string) => {
      // Future architecture hooks for memory poisoning/drift checks based on tool output
    }
  };
}

File: packages/aarm-sdk-python/aarm_sdk/frameworks/langchain.py
import json
from typing import Optional, Union, Dict, Any

from ..client import AarmClient
from ..types import RawRequest

class LangchainHooks:
    def __init__(self, client: AarmClient, session_id: str, intent: str, options: Optional[Dict[str, str]] = None):
        self.client = client
        self.session_id = session_id
        self.intent = intent
        self.options = options or {}

    def on_tool_start(self, tool_name: str, tool_input: Union[str, Dict[str, Any]]) -> Optional[Dict[str, Any]]:
        parameters: Dict[str, Any] = {}
        if isinstance(tool_input, str):
            try:
                parameters = json.loads(tool_input)
            except json.JSONDecodeError:
                parameters = {"input": tool_input}
        else:
            parameters = tool_input

        # Merge base options and build request
        business_unit = self.options.get('business_unit')
        application = self.options.get('application')
        principal_id = self.options.get('principal_id')

        request = RawRequest(
            session_id=self.session_id,
            intent=self.intent,
            user_prompt='Intercepted tool call',
            action_type=tool_name,
            parameters=parameters,
            business_unit=business_unit,
            application=application,
            principal_id=principal_id
        )

        mediation = self.client.mediate_action(request)

        if mediation.status == 'DENY':
            raise Exception(f"AARM Policy Violation: Execution of tool '{tool_name}' was blocked. Reason: {mediation.execution_status}")

        return mediation.receipt

    def on_tool_end(self, tool_name: str, output: str) -> None:
        # Future architecture hooks for memory poisoning/drift checks based on tool output
        pass

def create_langchain_hooks(client: AarmClient, session_id: str, intent: str, **options) -> LangchainHooks:
    """
    Creates callbacks/hooks intended to integrate seamlessly with LangChain's Tool execution flow.
    Note: Actual LangChain types are omitted to avoid steep dependency overhead in the base SDK.
    """
    return LangchainHooks(client, session_id, intent, options)
